README
Yeah, everybody puts nonsense they consider important in a place that's hard to miss, though they know that readers are skilled in skipping them. I'll do the same.

******
About:
******

This is an application meant to easen benchmarking of in-memory data transformations.
It started as a tool to simulate the way filesystems use compression, but its flexibility improved over time so much that testing other things works OK.
Aside from compression algorithms it features ciphers and hashes and could be repurposed for nearly anything.

*********
Features:
*********

* multithreading
* splitting data into blocks
* testing in any level of cache or in RAM
* overhead as low as 7 ticks/block on AMD64 and 11 on ARM

*****************
How does it work:
*****************

 1. It warms the CPU up with several iterations (-w) of memory shuffling.
 2. If job_size (-j) is not defined, it tries to come up with a good one.
 3. It splits the data into blocks (see -b switch).
 4. It creates encoding threads (-t).
 5. Each thread gets a number of blocks with a total size of at least job_size (-j). Note: Blocks are not split to smaller entities. When the entire file comes in just 1 block (which happens often with the defaults), it will be assigned to a single thread as a whole. The others won't get any work.
 6. Encoding of each block is repeated overhead_iterations number of times (-o). This is meant to reduce the overhead introduced by the benchmark itself. It's needed only for testing data that fits in L1 cache with the fastest transforms.
 7. Each thread encodes asigned blocks and asks for another job. Repeats until everything's compressed. If encoder is a compressor, block has to be reduced by it by at least sector_size of bytes (-m), otherwise it's left uncompressed.
 8. Each thread decodes __encoded__ blocks. Please note that these may be different from different codecs. If one failed to encode some block (i.e. compress by at least sector_size), it doesn't get to decode it. Decoding speed takes into account only blocks that were encoded successfully.
 9. Points 6 and 7 are repeated for at least small_iter_time ms. (-s), the runtimes are summed up. This is to compensate for clock inaccuracies.
10. Point 8 is repeated iters (-i) number of times, there's a minimum of runtimes taken. This is meant to reduce testing variability.

*****************
In-cache testing:
*****************

Testing in-cache is being performed by running encoder/decoder repeatadly on a piece of data that fits completely in your cache.
To do so, you need to specify a *small enough* block (with -b switch) and a fair number of overhead iterations (-o).
What is small enough? For ciphers / hashes, take slightly less than the size of your cache. For compressors - enough to contain both compressed and uncompressed data.

****************
Included codecs:
****************

Some codecs are in the default installation, some you have to enable with a compile-time option. See CMakeFiles.txt if you want them.

Checksums:
name                       | author                                      | version               | default? | source
--------------------------------------------------------------------------------------------------------------------
Blake224                   | multiple                                    | SHA3 Final, 64bit opt | yes      |
Blake256                   | multiple                                    | SHA3 Final, 64bit opt | yes      |
Blake384                   | multiple                                    | SHA3 Final, 64bit opt | yes      |
Blake512                   | multiple                                    | SHA3 Final, 64bit opt | yes      |
Blake2b                    | multiple                                    | 20130131              | yes      | https://blake2.net
Blake2bp                   | multiple                                    | 20130131              | yes      | https://blake2.net
Blake2s                    | multiple                                    | 20130131              | yes      | https://blake2.net
Blake2sp                   | multiple                                    | 20130131              | yes      | https://blake2.net
BlueMidnightWish224        | Danilo Gligoroski                           | SHA3 rnd 2, 64bit opt | no       |
BlueMidnightWish256        | Danilo Gligoroski                           | SHA3 rnd 2, 64bit opt | no       |
BlueMidnightWish384        | Danilo Gligoroski                           | SHA3 rnd 1, 64bit opt | no       |
BlueMidnightWish512        | Danilo Gligoroski                           | SHA3 rnd 1, 64bit opt | no       |
CityHash32                 | Geoff Pike, Jyrki Alakuijala                | 1.1.0                 | no       | https://code.google.com/p/cityhash
CityHash64                 | Geoff Pike, Jyrki Alakuijala                | 1.1.0                 | no       | https://code.google.com/p/cityhash
CityHash128                | Geoff Pike, Jyrki Alakuijala                | 1.1.0                 | no       | https://code.google.com/p/cityhash
CrapWow                    | Andrew@Team5150.com                         | 2012-06-07            | no       | www.team5150.com/~andrew/noncryptohashzoo
crypto++ - adler32         | Wei Dai                                     | 5.6.1                 | no       | www.cryptopp.com
crypto++ - crc32           | Wei Dai                                     | 5.6.1                 | no       | www.cryptopp.com
crypto++ - md5             | Colin Plumb, Wei Dai                        | 5.6.1                 | no       | www.cryptopp.com
crypto++ - sha224          | Steve Reid, Wei Dai                         | 5.6.1                 | no       | www.cryptopp.com
crypto++ - sha256          | Steve Reid, Wei Dai                         | 5.6.1                 | no       | www.cryptopp.com
crypto++ - sha384          | Steve Reid, Wei Dai                         | 5.6.1                 | no       | www.cryptopp.com
crypto++ - sha512          | Steve Reid, Wei Dai                         | 5.6.1                 | no       | www.cryptopp.com
CubeHash224                | Daniel J. Bernstein                         | SHA3 rnd 2, 64bit opt | no       |
CubeHash256                | Daniel J. Bernstein                         | SHA3 rnd 2, 64bit opt | no       |
CubeHash384                | Daniel J. Bernstein                         | SHA3 rnd 2, 64bit opt | no       |
CubeHash512                | Daniel J. Bernstein                         | SHA3 rnd 2, 64bit opt | no       |
Edon-R224                  | multiple                                    | v20                   | no       | www.item.ntnu.no/people/personalpages/fac/danilog/edon-r
Edon-R256                  | multiple                                    | v20                   | no       | www.item.ntnu.no/people/personalpages/fac/danilog/edon-r
Edon-R384                  | multiple                                    | v20                   | no       | www.item.ntnu.no/people/personalpages/fac/danilog/edon-r
Edon-R512                  | multiple                                    | v20                   | no       | www.item.ntnu.no/people/personalpages/fac/danilog/edon-r
FarmHash32                 | ?                                           | SVN r4                | yes      | https://code.google.com/p/farmhash
FarmHash64                 | ?                                           | SVN r4                | yes      | https://code.google.com/p/farmhash
FarmHash128                | ?                                           | SVN r4                | yes      | https://code.google.com/p/farmhash
fletcher2                  | ?? (used in ZFS)                            | 2010                  | no       |
fletcher4                  | ?? (used in ZFS)                            | 2010                  | no       |
FNV1a-Jesteress            | Georgi Marinov                              | 2012-06-09            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Mantis               | Georgi Marinov                              | 2012-06-09            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Meiyan               | Georgi Marinov                              | 2012-06-09            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Penumbra             | Georgi Marinov                              | 2013-06-16            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Tesla                | Georgi Marinov                              | v2                    | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Tesla3               | Georgi Marinov                              | 2013-03-13            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Yorikke              | Georgi Marinov                              | 2013-05-12            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-YoshitsuTRIAD        | Georgi Marinov                              | 2013-05-12            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-YoshimitsuTRIADiiXMM | Georgi Marinov                              | 2013-05-12            | yes      | www.sanmayce.com/Fastest_Hash/index.html
FNV1a-Yoshimura            | Georgi Marinov                              | 2013-03-13            | yes      | www.sanmayce.com/Fastest_Hash/index.html
Grøstl224                  | Soeren S. Thomsen, Krystian Matusiewicz     | SHA3 Final, 64bit opt | yes      |
Grøstl256                  | Soeren S. Thomsen, Krystian Matusiewicz     | SHA3 Final, 64bit opt | yes      |
Grøstl384                  | Soeren S. Thomsen, Krystian Matusiewicz     | SHA3 Final, 64bit opt | yes      |
Grøstl512                  | Soeren S. Thomsen, Krystian Matusiewicz     | SHA3 Final, 64bit opt | yes      |
JH224                      | Hongjun Wu                                  | SHA3 Final, 64bit opt | yes      |
JH256                      | Hongjun Wu                                  | SHA3 Final, 64bit opt | yes      |
JH384                      | Hongjun Wu                                  | SHA3 Final, 64bit opt | yes      |
JH512                      | Hongjun Wu                                  | SHA3 Final, 64bit opt | yes      |
Keccak224                  | multiple                                    | SHA3 Final, 64bit opt | yes      |
Keccak256                  | multiple                                    | SHA3 Final, 64bit opt | yes      |
Keccak384                  | multiple                                    | SHA3 Final, 64bit opt | yes      |
Keccak512                  | multiple                                    | SHA3 Final, 64bit opt | yes      |
murmur3_x86_32             | Austin Appleby                              | 2012-02-29            | yes      |
murmur3_x86_128            | Austin Appleby                              | 2012-02-29            | yes      |
murmur3_x64_128            | Austin Appleby                              | 2012-02-29            | yes      |
SipHash                    | Jean-Philippe Aumasson, Daniel J. Bernstein | reference             | yes      | https://131002.net/siphash
Skein224                   | Doug Whiting                                | SHA3 Final, 64bit opt | yes      |
Skein256                   | Doug Whiting                                | SHA3 Final, 64bit opt | yes      |
Skein384                   | Doug Whiting                                | SHA3 Final, 64bit opt | yes      |
Skein512                   | Doug Whiting                                | SHA3 Final, 64bit opt | yes      |
Skein1024                  | Doug Whiting                                | SHA3 Final, 64bit opt | yes      |
SpookyHash                 | Bob Jenkins                                 | V2 2012-08-05         | yes      | http://burtleburtle.net/bob/hash/spooky.html
SWIFFTX224                 | multiple                                    | SHA3 rnd 1, 64bit opt | no       |
SWIFFTX256                 | multiple                                    | SHA3 rnd 1, 64bit opt | no       |
SWIFFTX384                 | multiple                                    | SHA3 rnd 1, 64bit opt | no       |
SWIFFTX512                 | multiple                                    | SHA3 rnd 1, 64bit opt | no       |
uhash                      | Ted Krovetz                                 | 2007-04-17            | yes      | www.fastcrypto.com
vhash                      | Ted Krovetz, Wei Dai                        | 2007-04-17            | yes      | www.fastcrypto.com
umac                       | Ted Krovetz                                 | 2007-04-17            | yes      | www.fastcrypto.com
vmac                       | Ted Krovetz, Wei Dai                        | 2007-04-17            | yes      | www.fastcrypto.com
xxhash                     | Yann Collet                                 | r39                   | yes      | https://code.google.com/p/xxhash
xxhash64                   | Yann Collet                                 | r39                   | yes      | https://code.google.com/p/xxhash


Ciphers:
name                                | author                                  | version    | default? |
-------------------------------------------------------------------------------------------------------
AES128                              | Daniel J. Bernstein                     | little-4   | yes      |
AES256                              | Hongjun Wu                              | v1         | yes      |
ChaCha                              | Daniel J. Bernstein                     | 20080118   | yes      |
HC-128                              | ??                                      | 2007-01b   | yes      |
HC-256                              | ??                                      | 2007-01    | yes      |
LEX                                 | Alex Biryukov                           | v2         | yes      |
Rabbit                              | ??                                      | opt2       | yes      |
RC4                                 | Christophe De Cannière                  | 2005-08-21 | yes      |
Salsa20/8                           | Daniel J. Bernstein                     | merged     | yes      |
Salsa20/12                          | Daniel J. Bernstein                     | merged     | yes      |
Salsa20                             | Daniel J. Bernstein                     | merged     | yes      |
Snow 2.0                            | Patrik Ekdahl and Thomas Johansson      | fast       | yes      |
Sosemanuk                           | Thomas Pornin                           | 2005-04-26 | yes      |
Trivium                             | Christophe De Cannière                  | 2006-02-23 | yes      |

Compressors:
name                                | author                                           | version             | default? | source
--------------------------------------------------------------------------------------------------------------------------------
7z - deflate                        | Igor Pavlov                                      | 9.20                | yes      | http://7-zip.org
7z - deflate64                      | Igor Pavlov                                      | 9.20                | yes      | http://7-zip.org
ar                                  | Haruhiko Okumura, Kerwin F. Medina, SanguineRose | 2013-06-03          | no       | https://github.com/Sanguinarious/lzh-improved
Basic Compression Library - huffman | Marcus Geelnard                                  | 1.2.0               | no       | http://bcl.comli.eu
Basic Compression Library - lz      | Marcus Geelnard                                  | 1.2.0               | no       | http://bcl.comli.eu
Basic Compression Library - lzfast  | Marcus Geelnard                                  | 1.2.0               | no       | http://bcl.comli.eu
Basic Compression Library - rle     | Marcus Geelnard                                  | 1.2.0               | no       | http://bcl.comli.eu
blosc                               | Francesc Alted                                   | 1.5.2               | no       | http://blosc.pytables.org/trac
BriefLZ                             | Joergen Ibsen                                    | 1.0.5               | no       | http://ibsensoftware.com/download.html
Brotli                              | Jyrki Alakuijala, Zoltan Szabadka                | 2015-02-01          | no       | https://github.com/google/brotli
bzip2                               | Julian Seward                                    | 1.0.6               | yes      | http://bzip.org
crush                               | Ilia Muraviev                                    | 0.0.1               | yes      | http://encode.ru/threads/1289-CRUSH-0-01-is-here!
crypto++ - deflate                  | Wei Dai                                          | 5.6.1               | yes      | www.cryptopp.com
Density                             | Guillaume Voirin                                 | 2015-02-02          | yes      | https://github.com/centaurean/density
Doboz                               | Attila T. Áfra                                   | 2011-03-19          | no       | https://bitbucket.org/attila_afra/doboz/downloads
FastLZ                              | Ariya Hidayat                                    | 0.1.0               | no       | http://fastlz.org
FSE                                 | Yann Collet                                      | 2014-04-07          | no       | https://github.com/Cyan4973/FiniteStateEntropy
gipfeli                             | Rasto Lenhardt, Jyrki Alakuijala                 | 2011-10-19          | no       | https://code.google.com/p/gipfeli
Halibut - deflate                   | Simon Tatham                                     | SVN r9550           | no       | http://www.chiark.greenend.org.uk/~sgtatham/halibut
lodepng - deflate                   | Lode Vandevenne                                  | 20120729            | no       | http://lodev.org/lodepng
lrrle                               | m^2                                              | 0                   | yes      |
LZ4                                 | Yann Collet                                      | r127                | yes      | https://code.google.com/p/lz4
LZ4hc                               | Yann Collet                                      | r127                | yes      | https://code.google.com/p/lz4
LZF                                 | Marc Alexander Lehmann                           | 3.6                 | yes      | http://oldhome.schmorp.de/marc/liblzf.html
LZFX                                | Andrew Collette                                  | r16                 | no       | https://code.google.com/p/lzfx
LZG                                 | Marcus Geelnard                                  | 1.0.6               | no       | http://liblzg.bitsnbites.eu
LZHAM                               | Richard Geldreich, Jr.                           | 2015-01-31          | no       | https://github.com/richgel999/lzham_codec
LZJB                                | Jeff Bonwick                                     | FreeBSD r263244     | no       |
mmini - lzl                         | Adam Ierymenko                                   | 2010                | no       | https://code.google.com/p/mmini
LZMAT                               | Vitaly Evseenko                                  | 1.1                 | yes      | www.matcode.com/lzmat.htm
LZO                                 | Markus Franz Xaver Johannes Oberhumer            | 2.08                | yes      | www.oberhumer.com/opensource/lzo
LZP_ds                              | Dmitry Shkarin                                   | v0                  | no       | http://www.ctxmodel.net/files/LZP-DS_v0.rar
LZSS                                | Ilia Muraviev                                    | 2008-07-31          | no       | http://encode.ru/threads/143-LZSS-v0-01-is-here!
LZSS                                | Nobuo Ito                                        | 1.0                 | no       |
LZV1                                | Hermann Vogt                                     | 0.5                 | no       | http://encode.ru/threads/1661-LZWC-A-fast-tree-based-LZW-compressor
lzx_compress                        | Matthew T. Russotto                              | 2005-07-06          | no       | http://www.russotto.net/chm/
miniz                               | Richard Geldreich, Jr.                           | 1.11                | yes      | https://code.google.com/p/miniz
Nakamichi                           | Georgi Marinov                                   | Aratama             | yes      | http://www.sanmayce.com/Nakamichi/
Nakamichi                           | Georgi Marinov                                   | Butsuhira           | yes      | http://www.sanmayce.com/Nakamichi/
Nakamichi                           | Georgi Marinov                                   | ButsuhiraBranchless | yes      | http://www.sanmayce.com/Nakamichi/

src/CHANGELOG
0.15
[+] added FarmHash SVN r4
[+] added Intel's fork of zlib
[+] added memcpy, memmove, memset codecs
[+] added crush 0.0.1
[+] added zling 20140324
[+] added Nakamichi in many variants
[+] added PostgreSQL's LZ 9.3.4
[+] added FSE 2014-04-07
[+] added Nobuo Ito's LZSS 1.0
[+] added Dmitry Shkarin's LZP v0
[+] added lrrle v0
[+] added zstd 2015-01-31
[+] added Brotli 2015-02-01
[+] added wfLZ r10
[+] added density 2015-02-02
[~] updated blosc to 1.5.2
[~] updated lz4 to r127
[~] updated LZJB to the latest FreeBSD version
[~] updated tornado to 0.6a
[~] updated xxhash to r39
[~] updated LZO to 2.08
[~] updated LZHAM to 2015-01-31
[~] updated gipfeli to 2014-06-27
[~] made CityHash optional as it's superseeded by FarmHash
[-] removed xxhash256, it's dangerously weak
[!] fixed integer underflow
[!] fixed a segfault with nop codec
[!] fixed an error in decompression efficiency calculations

0.14.2
[+] added FNV1a-penumbra
[!] fixed compilation issues on some Linuces

0.14.1
[+] added byte-swapping transforms (bswap16, bswap32, bswap64)
[!] added missing zconf.h
[!] minor fixes

0.14
[+] added z3lib 1.3
[+] added CityHash32
[+] added 128-bit variants of murmur hash
[+] added support for safe lz4 decoder. Use LZ4 for the fast one (like before) and LZ4safe for the safe one
[~] removed bcl-lz from the list of all codecs. It's so extremely slow it was disturbing testing. And lzfast is still there.
[~] the number of small_iters is not fixed; duration is
[~] updated CityHash64/128 to 1.1.0
[~] updated zlib to 1.2.8
[~] updated snappy to 1.1.0
[~] updated lzo to 2.06
[~] updated blosc to 1.2.3
[~] updated zopfli to 1.0.0
[~] updated SpookyHash to V2 2012-08-05
[~] updated LZ4 to r97
[~] cleanup
[!] restoring default console colour would fail on some systems
[!] several minor bugs

0.13.4
[+] added umac, vmac
[+] added ar
[~] cleanup
[~] improved printing of results
[!] added uhash, vhash, umac, vmac to the list of checksums
[!] Hopefully fixed compilation issues on some Unices 

0.13.3
[+] added uhash, vhash
[!] disabled FNV1a-YosimitsuTriadiiXMM unless SSL2 is explicitly enabled

0.13.2
[+] added more Sanmayce's hashes
[~] updated LZ4 to r95
[~] updated xxhash to r29
[-] removed LZ4bz because I don't want to handle the burden of maintaining the patch since Yann just totally changed packaging of LZ4
[!] portability issues

0.13.1
[~] minor cleanup and refactoring
[~] updated Blake2 to 20130131
[!] blake2s, blake2sp have 256-bit digests
[!] compilation errors on some platforms

0.13
[+] support for pipelining compression->encryption->checksumming (and other pipelines too)
[+] added a couple of SHA3 candidate hashes
[+] added gipfeli
[+] added siphash
[+] added blake2
[+] added LZ4bz - an incompatible mod of LZ4 proposed by Bulat Ziganshin
[+] added mmini (only LZ77 for now)
[+] added zopfli
[+] added new Sanmayce's hashes
[~] updated LZ4 to r85, now I use LZ4_compress_limitedOutput function
[~] removed LZ4 option to use more memory, it was too hard to maintain
[~] updated xxhash to r5
[~] updated lodepng to 20120729
[~] cleanup, minor fixes and refactoring
[~] eclipse integration
[~] initial doxygen integration
[~] fossil integration
[!] updated Tornado to the latest mercurial version to fix a clang issue
[!] numerous portability issues
[!] memory leak in zlib and miniz handlers
[!] insignificant memory leak in the main testing function
[!] md5 has 128 bits, not 160

0.12
[+] updated Shrinker to r6
[+] updated fnv1a-Tesla to v2
[+] updated xxhash256 to v1
[+] added ECrypt ciphers

0.11
[+] added xxhash
[+] added Sanmayce's hashes
[+] added SpookyHash
[+] added MurmurHash
[+] added CityHash
[+] added crypto++'s hashes
[+] added zfs' fletcher2 and fletcher4 hashes
[+] added CrapWow hash
[+] refactoring to enable in-place transforms (i.e. ciphers, hashes)
[~] always use aligned buffers
[~] removed bzip2 from the default installation (it doesn't have any real competition in FSBench ATM)
[~] removed shrinker from the default installation (doesn't work on my linux)
[~] added crypto++ to the default build
[~] use crc instead of memcmp to detect incorrect (de)compression
[~] reworked buffer management
[~] code cleanup

0.10
[+] added QuickLZ zip
[+] added Halibut's deflate
[+] colourful output
[~] updated zlib to 1.2.7
[~] updated LZ4 (and LZ4hc that was merged with it) to r66
[~] removed LZ4 modes that used LZ4_COMPRESSMIN of 4096, it didn't really work different from 1
[~] code cleanup
[!] compilation errors on some systems

0.10h
[+] now you can combine encoder and decoder from different libraries, i.e. miniz encoder with zlib decoder
[+] you can now select which codecs to compile
[+] added 7-zip's deflate and deflate64
[+] added crypto++'s deflate
[+] added tinf
[+] added lodepng
[+] added scz
[+] added Basic Compression Library (bcl)
[+] added bzip2
[+] added an efficiency metric to the output: (input_size-compressed_size)/time
[~] disabled compilation of many codecs in the default configuration, edit CMakeLists.txt to enable them
[~] major revamp of output formatting to allow longer codec names
[~] renamed lzx to lzx_compress to make the name unambigous
[~] made zlib / miniz use deflate instead of zlib. They may work slightly faster because of dropped checksum calculations.
[~] because of portability issues, LZHAM is now optional
[~] use memcpy instead of LZ4 to determine the number of small iters; now the code takes number of threads into account too
[~] code cleanup
[!] fixes to enable compile-time removal of codecs
[!] fixed a miniz compilation issue on mingw32 (it tried to use __forceinline that didn't work there)
[!] Yappy decompression produced incorrect output without SSL

0.10g
[+] added blosc
[~] better warnings
[!] GBytes were missing from the list of units

0.10f
[+] added protection from codecs that report too small output size
[~] updated Shrinker to r5

src/codecs.hpp
/**
 * Written by m^2.
 * You can consider the code to be public domain.
 * If your country doesn't recognize author's right to relieve themselves of copyright,
 * you can use it under the terms of WTFPL version 2.0 or later.
 *
 *
 * To add a codec:
 * 1. Implement a Codec class for it or - if applicable - functions required by one of abstract ones
 * 2. Add sources to CMakeFiles.txt
 * 3. Add codec version to codecs.hpp
 * 4. Add codec to all_compressors / all_checksums / all_ciphers in codecs.cpp
 * 5. Update readme
 * 6. Update changelog
 * 7. Add sources to SCM
 */
#ifndef CODECS_HPP_BhjgkfG8
#define CODECS_HPP_BhjgkfG8

#include <list>
#include <exception>
#include <string>
#include <vector>

// some codecs have rough bounds checking
// to overcome the problem I allocate some extra data
#define OVERRUN_PAD 32

// in    - input buffer
// isize - input data size
// out   - output buffer
// osize - desired output size. Codec can write up to max_encoded_size(in), 
//         but if it fails to compress to osize or less, 
//         it can quit trying before the end of data
//         it's warranted that the actual buffer size is not smaller than max_encoded_size(isize)
//         and that actual size of the input buffer is just as big (which matters i.e. for in_place codecs)
typedef size_t (*encoder_t)(char * in, size_t isize, char * out, size_t osize, void * args);
// in    - input buffer
// isize - input data size
// out   - output buffer
// osize - output data size = decoded size
typedef size_t (*decoder_t)(char * in, size_t isize, char * out, size_t osize, void * args);
typedef size_t (*max_encoded_size_t)(size_t input_size);

// When some codec can warrant no size blowup (and i.e. error instead), this function can be used
// Still, I reserve a couple of bytes for buffer overruns, some codecs need it and it's not worth introducing another function for them
inline size_t no_blowup(size_t input_size)
{
    return input_size + OVERRUN_PAD;
}

// should be returned by encoder_t / decoder_t when something breaks (i.e. there's a bug or meeting osize requirements is impossible)
#define CODING_ERROR 0

/**
 * @brief A base class for all encoders / decoders
 */
struct Codec
{
    struct InvalidParams: std::exception
    {
        virtual const char * what() const throw ();
        InvalidParams(const std::string & message);
        ~InvalidParams() throw ()
        {
        }
    private:
        const std::string message;
    };
    enum transform_type
    {
        in_place, // takes an input buffer and modifies it in place
        moving,  // takes 2 buffers, input and output
        buffered,  // takes 2 buffers, input/output and a buffer
        selective // takes 2 buffers and puts data in whichever it prefers, returning information about its destination
    };

    const std::string name;
    const std::string version;
    std::string args;

    encoder_t encoder;
    decoder_t decoder;
    virtual size_t max_encoded_size(size_t input_size);

    transform_type encode_transform_type;
    transform_type decode_transform_type;

    bool can_be_skipped; //when a compressor fails to reduce size, the data can be left uncompressed and decompression can be skipped
                         //this doesn't hold for checksums, they have to be calculated even though they don't save space ;)

    virtual void ** eparams(); // parmeters to be sent to compressors
    virtual void ** dparams(); // parmeters to be sent to decompressors 

    /**
     * Creates a codec
     * @throw InvalidParams Dude, something's wrong with your command line. 
     */
    Codec(const std::string & name,
          const std::string & version,
          encoder_t encoder,
          decoder_t decoder,
          max_encoded_size_t max_size = _max_compressed_size,
          transform_type encode_transform_type = moving,
          transform_type decode_transform_type = moving,
          bool can_be_skipped = true);

    // returns a help message
    virtual std::string help() const;
    // returns codec name, version, args and whatever other information the codec would like to say about itself
    virtual std::string introduction() const;

    // (de)initialisation functions
    // set the following fields:
    // - args
    // - encoder
    // - decoder
    // - max_compressed_size
    // - (subclasses) whatever else do they need.
    // - encode_transform_type
    // - decode_transform_type
    // the following are set by constructor:
    // - name
    // - version
    // - can_be_skipped
    virtual void init(const std::string & args,
                      unsigned threads_no,
                      size_t isize,
                      bool init_compressor = true,
                      bool init_decompressor = true);
    virtual void cleanup();
    virtual ~Codec()
    {
    }

protected:
    // the default, fairly-safe function to be used with codecs that don't have their own
    static size_t _max_compressed_size(size_t input_size);
    max_encoded_size_t real_max_compressed_size;

};

/**
 * @brief Stores a codec and parameters that are supposed to be used by it
 */
struct CodecWithParams
{
    Codec & codec;
    const std::string params;
    CodecWithParams(Codec & codec, const std::string & params) :
            codec(codec), params(params)
    {
    }
    CodecWithParams operator=(const CodecWithParams & src)
    {
        return CodecWithParams(src.codec, src.params);
    }
};

extern std::list<Codec*> CODECS;
extern std::list<CodecWithParams> DEFAULT_CODECS;
extern std::list<CodecWithParams> FAST_COMPRESSORS;
extern std::list<CodecWithParams> ALL_COMPRESSORS;
extern std::list<CodecWithParams> ALL_CHECKSUMS;
extern std::list<CodecWithParams> ALL_CIPHERS;
extern std::list<CodecWithParams> ALL_OTHERS;

Codec * find_codec(const std::string & name);

// TODO: each codec should have a version number by itself,
// so there's one place less to change when adding / updating 
#define _7z_VERSION             "9.20"
#define _AR_VERSION             "2013-06-03"
#define _BCL_VERSION            "1.2.0"
#define _BLAKE2_VERSION         "20121223"
#define _BLOSC_VERSION          "1.2.3"
#define _BLZ_VERSION            "1.0.5"
#define _BROTLI_VERSION         "2015-02-01"
#define _BZ2_VERSION            "1.0.6"
#define _CITYHASH_VERSION       "1.1.0"
#define _CRAPWOW_VERSION        "2012-06-07"
#define _CRUSH_VERSION          "0.0.1"
#define _CRYPTOPP_VERSION       "5.6.1"
#define _DENSITY_VERSION        "0.12.5 beta"
#define _DOBOZ_VERSION          "2011-03-19"
#define _ECRYPT_VERSION         "2012-06-10"
#define _EDON_R_VERSION         "v20"
#define _FARMHASH_VERSION       "SVN r4"
#define _FASTCRYPTO_VERSION     "2007-04-17"
#define _FASTLZ_VERSION         "0.1.0"
#define _FSE_VERSION            "2014-04-07"
#define _GIPFELI_VERSION        "2014-06-27"
#define _HALIBUT_VERSION        "SVN r9550"
#define _KECCAK_VERSION         "3.2"
#define _LODEPNG_VERSION        "20120729"
#define _LRRLE_VERSION          "0"
#define _LZ4_VERSION            "r129"
#define _LZF_VERSION            "3.6"
#define _LZFX_VERSION           "r16"
#define _LZG_VERSION            "1.0.6"
#define _LZHAM_VERSION          "2015-01-31"
#define _LZMAT_VERSION          "1.1"
#define _LZO_VERSION            "2.08"
#define _LZP_DS_VERSION         "v0"
#define _LZSSIM_VERSION         "2008-07-31"
#define _LZV1_VERSION           "0.5"
#define _LZWC_VERSION           "0.4"
#define _LZX_VERSION            "2005-07-06"
#define _MINIZ_VERSION          "1.11"
#define _MMINI_VERSION          "2012-12-23"
#define _MURMUR_VERSION         "2012-02-29"
#define _NAKAMICHI_VERSION(name) STRIGIFY(name)
#define _NAKAMICHI_M_VERSION    "M2.0"
#define _NOBUO_ITO_LZSS_VERSION "1.0"
#define _NRV_VERSION            "1.03"
#define _PG_LZ_VERSION          "9.3.4"
#define _QLZ_VERSION            "1.5.1b6"
#define _QLZZIP_VERSION         "0.4"
#define _RLE64_VERSION          "R3.00"
#define _SANMAYCE_VERSION       "2013-06-16"
#define _SIPHASH_VERSION        "reference"
#define _SCZ_VERSION            "2008-11-25"
#define _SHA3_VERSION           "SHA3 Final 64bit opt"
#define _SHA3_RND1_VERSION      "SHA3 rnd 1 64bit opt"
#define _SHA3_RND2_VERSION      "SHA3 rnd 2 64bit opt"
#define _SHRINKER_VERSION       "r6"
#define _SNAPPY_VERSION         "1.1.0"
#define _SPOOKY_VERSION         "V2 2012-08-05"
#define _TINF_VERSION           "1.00"
#define _TOR_VERSION            "0.6a"
#define _WFLZ_VERSION           "r10"
#define _XXHASH_VERSION         "r39"
#define _XXHASH256_VERSION      "1"
#define _YAPPY_VERSION          "v2"
#define _ZFS_VERSION            "FreeBSD r263244"
#define _Z3LIB_VERSION          "1.3"
#define _ZLIB_VERSION           "1.2.8"
#define _ZLIB_INTEL_VERSION     "2014-06-16 Intel"
#define _ZLING_VERSION          "20140324"
#define _ZOPFLI_VERSION         "1.0.0"
#define _ZSTD_VERSION           "2015-01-31"
#endif // CODECS_HPP_BhjgkfG8
